### DISCUSSION TITLE
armhf: Easier way to build against glibc > 2.17? (or lower dependency glibc requirements)

### DISCUSSION THREAD

**Discussion by Trayshar** [cross-rs/cross#875] [Q&A]
Heya! Im building my project, which depends on `libudev-dev` against `arm-unknown-linux-gnueabihf`(which compiles against glibc 2.17). After some tweaking (see #874) it managed to nearly compile, except I get 
`/lib/arm-linux-gnueabihf/libudev.so: undefined reference to getrandom@GLIBC_2.25'.`
Building my own base image using crosstool-ng (see #734 for instructions) against glibc 2.27 works, but this was harder than expected since there seems to be no documentation outside the commit message in #734. Also, it's not good for shared projects, since my other colleagues don't have my "homemade" base image to compile with (same goes for CI) and compilation of this custom image took around ~2h on my machine.

Is there a way to either get a image like this automatically or lower the glibc requirements of libudev-dev?

**Comment by Alexhuszagh** [comment]
We have documentation elsewhere but we should probably improve it/move it to cross itself, probably the Wiki. For example, in [cross-toolchains](https://github.com/cross-rs/cross-toolchains#configure) we have:

```markdown
The config files are configured via [crosstool-ng/configure-all.sh](https://github.com/cross-rs/cross-toolchains/blob/main/crosstool-ng/configure-all.sh), which may be customized by the following environment variables:
- `GCC_VERSION`: The GCC version (default 8.3.0)
- `GLIBC_VERSION`: The glibc version (default 2.17)
- `LINUX_VERSION`: The Linux version (default 4.19.21)
```

The same thing is actually true for [cross-rs](https://github.com/cross-rs/cross/blob/main/crosstool-ng/configure.sh). Maybe we could add it to the Wiki? Does that sound good?

So in this case, you should be able to auto-generate the config file you want with `GLIBC_VERSION=2.27 crosstool-ng/configure.sh`. We really should move this away from a bash script however and let you use it for individual targets.

**Reply by Trayshar** [reply]
Yeah, I generated my image using `GLIBC_VERSION=2.27 crosstool-ng/configure.sh`, that works. Having these options for individual targets would be awesome! Also, I wish there would be an easier way to do that than 

1. Cloning the cross-rs repository
2. Running the aforementioned command
3. Building and tagging the image
4. Using the custom image
5. Giving these instructions to anyone else who wishes to compile my project

Sadly, I'm not familiar enough with the internals of cross to make a good suggestion here. Improving the documentation is a good idea in general and seems like a good medium-term solution.

**Reply by Alexhuszagh** [reply]
@Trayshar It would be somewhat difficult, since it would require maintaining multiple images of different versions, but:
1. Improving configure of specific targets using `cross-rs` by switching it to a Rust command.
2. Improving the documentation.

Due to the reasons mentioned in the [comments](https://github.com/cross-rs/cross/issues/874#issuecomment-1167375132) of your issue, I don't think bumping the glibc version is a great idea, due to backwards compatibility concerns. I think documenting `vcpkg` or `conan` use is a better idea. Sound good?

**Reply by Alexhuszagh** [reply]
Specifically, for `vcpkg`, these are the scripts I use to install it, which would then allow you to use it as part of your build scripts:

[vcpkg.sh.in](https://github.com/Alexhuszagh/xcross/blob/35dcdda18aec7bea39ca3a102b0ef3c827864460/docker/vcpkg.sh.in)
[vcpkg-triplet.sh.in](https://github.com/Alexhuszagh/xcross/blob/507458c6fd4168a98d38ddc11e272c6e11260113/docker/vcpkg-triplet.sh.in)

Please note these aren't configured, but can easily be so (everything is unlicensed, so use it as you wish: no attribution is required). The configured scripts are as follows:

**vcpkg.sh**

```bash
#!/bin/bash
#
# Install vcpkg package management system.

set -ex

# There's a minor, well-known issue with wasm and vcpkg.
# vcpkg runs in script mode for the CMake files, meaning
# the `CMAKE_EXECUTABLE_SUFFIX` is not set, so all binaries
# will not have the `.js` suffix (although the wasm ones will).
#   https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#useful-implementation-notes

export DEBIAN_FRONTEND="noninteractive"

# Install dependencies. We store the installed
# dependencies so we don't accidentally delete
# necessary files, and we get rid of everything
# that was only required for the build.
before_installed=($(apt -qq list --installed 2>/dev/null | cut -d '/' -f 1))
apt-get update
apt-get install --assume-yes --no-install-recommends \
    curl \
    g++ \
    git \
    ca-certificates \
    jq \
    wget \
    unzip
after_installed=($(apt -qq list --installed 2>/dev/null | cut -d '/' -f 1))

# Calculate the packages we need to remove later.
diff=()
for i in "${after_installed[@]}"; do
    skip=
    for j in "${before_installed[@]}"; do
        [[ $i == $j ]] && { skip=1; break; }
    done
    [[ -n $skip ]] || diff+=("$i")
done
declare -p diff

# Download our latest release.
url="https://api.github.com/repos/microsoft/vcpkg/releases/latest"
release=$(curl "$url" --silent | jq -r '.zipball_url')
wget "$release" -O vcpkg.zip
mkdir -p "/opt"/vcpkg/
unzip vcpkg.zip -d "/opt"/vcpkg/
rm -f vcpkg.zip
subdir=$(ls "/opt"/vcpkg)
shopt -s dotglob
mv "/opt"/vcpkg/"$subdir"/* "/opt"/vcpkg
shopt -u dotglob
rmdir "/opt"/vcpkg/"$subdir"

# Bootstrap the build, and remove unnecessary extras.
"/opt"/vcpkg/bootstrap-vcpkg.sh

# Change our permissions, since we need to be able to write as a user.
# Use the user ID, not the actual username, since the user hasn't yet
# been created.
chmod -R +660 "/opt"/vcpkg
chown -R 1000:1000 "/opt"/vcpkg

# Cleanup
# Even though we didn't install GCC, it doesn't get autoremoved,
# which takes up a lot of space.
apt-get remove --purge --assume-yes "${diff[@]}"
apt-get autoremove --assume-yes
```

**vcpkg-triple.sh**

```bash
#!/bin/bash
#
# Create the vcpkg wrappers to add the triplet and other
# configurations out-of-the-box.

set -ex

# Check required environment variables.
if [ "$PROCESSOR" = "" ]; then
    echo 'Must set the target processor via `$PROCESSOR`, quitting.'
    exit 1
fi
if [ "$SYSTEM" = "" ]; then
    echo 'Must set the target operating system via `$SYSTEM`, quitting.'
    exit 1
fi
if [ "$TRIPLE" = "" ]; then
    echo 'Must set the target triplet via `$TRIPLE`, quitting.'
    exit 1
fi
if [ "$LINKAGE" = "" ]; then
    echo 'Must set the target triplet via `$LINKAGE`, quitting.'
    echo 'Valid values are `static` and `dynamic`.'
    exit 1
fi

export DEBIAN_FRONTEND="noninteractive"

# Install dependencies we need after to use vcpkg.
apt-get update
apt-get install --assume-yes --no-install-recommends \
    curl \
    pkg-config \
    tar \
    unzip \
    zip

# Create our custom triplet, as a drop-in replacement for a community one.
triplet="set(VCPKG_TARGET_ARCHITECTURE $PROCESSOR)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE $LINKAGE)
set(VCPKG_CMAKE_SYSTEM_NAME $SYSTEM)

# The toolchain file can be included multiple times, which can
# cause linkage errors with installing libraries with dependencies.
if(NOT DEFINED VCPKG_CHAINLOAD_TOOLCHAIN_FILE)
    set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE /toolchains/toolchain.cmake)
endif()

if(NOT CMAKE_HOST_SYSTEM_PROCESSOR)
    execute_process(COMMAND \"uname\" \"-m\" OUTPUT_VARIABLE CMAKE_HOST_SYSTEM_PROCESSOR OUTPUT_STRIP_TRAILING_WHITESPACE)
endif()"
# Need to add items for MinGW and other systems.
if [ "$SYSTEM" = "MinGW" ]; then
    triplet="$triplet
set(VCPKG_ENV_PASSTHROUGH PATH)"
    if [ "$LINKAGE" = "dynamic" ]; then
        triplet="$triplet
set(VCPKG_POLICY_DLLS_WITHOUT_LIBS enabled)"
    fi
elif [ "$SYSTEM" = "Emscripten" ]; then
    triplet="$triplet
set(VCPKG_ENV_PASSTHROUGH EMSDK PATH)"
fi
triplet_file="/opt"/vcpkg/triplets/community/"$TRIPLE"-xcross.cmake
echo "$triplet" > "$triplet_file"
chmod +660 "$triplet_file"
chown -R cross:cross "$triplet_file"

# Create an alias for vcpkg.
echo '#!/bin/bash
# Wrapper script for vcpkg to ensure we use the correct triplet.
# Note that the triplet can always be provided: vcpkg does not care.

"/opt"/vcpkg/vcpkg "$@" --triplet $(cat /env/triplet)-xcross' > "/opt/bin/"/vcpkg
chmod +x "/opt/bin/"/vcpkg

# Create an environment variable containing the triplet.
# Allows us to call the correct triplet whenever we invoke vcpkg.
echo "$TRIPLE" > /env/triplet
```

You can then add them to the Dockerfile as follows:

```Dockerfile
FROM {repository}:{tag}

COPY ["docker/vcpkg.sh", "/"]
RUN /vcpkg.sh

COPY ["docker/vcpkg-triplet.sh", "/"]
RUN LINKAGE=static PROCESSOR=arm SYSTEM=Linux TRIPLE=arm-unknown-linux-gnueabihf /vcpkg-triplet.sh
```

You might want to install `conan` and maybe `meson` as well in these images, just so you have a complete C++ toolchain setup, making sure you can build almost all C++ packages from source. We might want to document this.

**Reply by Trayshar** [reply]
> @Trayshar It would be somewhat difficult, since it would require maintaining multiple images of different versions, but:
> 
> 1. Improving configure of specific targets using `cross-rs` by switching it to a Rust command.
> 2. Improving the documentation.
> 
> Due to the reasons mentioned in the [comments](https://github.com/cross-rs/cross/issues/874#issuecomment-1167375132) of your issue, I don't think bumping the glibc version is a great idea, due to backwards compatibility concerns. I think documenting `vcpkg` or `conan` use is a better idea. Sound good?

Having more control over the builds via commands would be neat. Documentation too. Sounds good to me.

>Specifically, for vcpkg, these are the scripts I use to install it, which would then allow you to use it as part of your build scripts:
>
>[vcpkg.sh.in](https://github.com/Alexhuszagh/xcross/blob/35dcdda18aec7bea39ca3a102b0ef3c827864460/docker/vcpkg.sh.in)
>[vcpkg-triplet.sh.in](https://github.com/Alexhuszagh/xcross/blob/507458c6fd4168a98d38ddc11e272c6e11260113/docker/vcpkg-triplet.sh.in)
>
>Please note these aren't configured, but can easily be so (everything is unlicensed, so use it as you wish: no attribution is required). The configured scripts are as follows:
> ...

I really appreciate your effort, thank you very much! To my suprise my build for amrhf that's linked against armv7-libudev still seems to work just fine. If it breaks I likely just try to compile my project on a native platform (I only care about raspberry  pi 1 and zero on armhf, and luckily got a bunch of those) since cross compilation for this target gets messy. I really wish there would an easier solution to this than having a bunch of scripts I do not know how to maintain or understand in my build setup (cross advertises itself as *"Zero setup” cross compilation*, after all). I do not have enough experience in this field to come up with a better solution tho (and sadly lack the time right now to get into it). More documentation on this would be great!

**Reply by Alexhuszagh** [reply]
It's pretty-much zero-setup (besides rustup + docker) for native Rust libraries, or those with vendored C++ dependencies, but yeah, for some odd cases it can be much more complex.

> To my suprise my build for amrhf that's linked against armv7-libudev still seems to work just fine.

This actually surprised me too, since I was fairly certain before I checked the architecture notes that it was ARMv7.

>  I do not have enough experience in this field to come up with a better solution tho (and sadly lack the time right now to get into it). More documentation on this would be great!

No worries, I'm lucky to have been a C++ developer before I started using Rust more, so I happen to be familiar with this. C++ packaging is an absolute mess, and I'm glad I can help with this.

**Reply by Trayshar** [reply]
> It's pretty-much zero-setup (besides rustup + docker) for native Rust libraries, or those with vendored C++ dependencies, but yeah, for some odd cases it can be much more complex.

Yeah. I'll continue using cross for the forseeable future. It's a great tool.

> > To my suprise my build for amrhf that's linked against armv7-libudev still seems to work just fine.
> 
> This actually surprised me too, since I was fairly certain before I checked the architecture notes that it was ARMv7.

Oh, it is ARMv7 alright, I checked it too. I compile using a armv6 toolchain, getting an amrv6 binary, but linking against an armv7 library in the process. It's linked dynamically tho, so perhaps this is why? Also, I haven't tested every hardware combination my project can run on yet, doing so in the coming weeks. Maybe something related to this will pop up.

> > I do not have enough experience in this field to come up with a better solution tho (and sadly lack the time right now to get into it). More documentation on this would be great!
> 
> No worries, I'm lucky to have been a C++ developer before I started using Rust more, so I happen to be familiar with this. C++ packaging is an absolute mess, and I'm glad I can help with this.

Thank you for your service ^^

**Reply by Alexhuszagh** [reply]
@Trayshar  I've updated the Wiki extensively and added a sample project for how to do this. This works out-of-the-box for ARMv6 targets. [vcpkg, Meson, and Conan](https://github.com/cross-rs/cross/wiki/Recipes#vcpkg-meson-and-conan).

**Reply by Trayshar** [reply]
> @Trayshar I've updated the Wiki extensively and added a sample project for how to do this. This works out-of-the-box for ARMv6 targets. [vcpkg, Meson, and Conan](https://github.com/cross-rs/cross/wiki/Recipes#vcpkg-meson-and-conan).

Thank you very much for that. I can't seem to find the ARMv6 example tho, only dockerfiles and configuration for ARMv7 and ARMv8. Am I looking in the right [place](https://github.com/cross-rs/wiki_assets/tree/main/Recipes/vcpkg)?

**Reply by Alexhuszagh** [reply]
> > @Trayshar I've updated the Wiki extensively and added a sample project for how to do this. This works out-of-the-box for ARMv6 targets. [vcpkg, Meson, and Conan](https://github.com/cross-rs/cross/wiki/Recipes#vcpkg-meson-and-conan).
> 
> Thank you very much for that. I can't seem to find the ARMv6 example tho, only dockerfiles and configuration for ARMv7 and ARMv8. Am I looking in the right [place](https://github.com/cross-rs/wiki_assets/tree/main/Recipes/vcpkg)?

Yep, just change this though to the ARMv6 image and it will work. This is because the toolchain compiles for ARMv6, so all this code will work (as long as you modify it for the compiler version in the README, which I believe is GCC 8.3.0):

```Dockerfile

# These are ordered from longest to shortest.
ARG VCKPKG_PROCESSOR=armhf
ARG PROCESSOR=arm
ARG SYSTEM=Linux
ARG TRIPLE=armelhf-unknown-linux-gnueabi
ARG QEMU=qemu-arm
ARG SYSROOT=/usr/arm-linux-gnueabihf
ARG CC=arm-linux-gnueabihf-gcc
ARG CXX=arm-linux-gnueabihf-g++
ARG AR=arm-linux-gnueabihf-ar
ARG STRIP=arm-linux-gnueabihf-strip

RUN mkdir -p /opt/cross/bin

COPY vcpkg-triple.sh /
RUN LINKAGE=static /vcpkg-triple.sh

COPY meson.sh /
RUN CPU_FAMILY=arm /meson.sh

COPY conan.sh /
RUN COMPILER=gcc COMPILER_VERSION="9" /conan.sh
```

**Reply by Alexhuszagh** [reply]
Also, the image build times are super quick (order of like 3-5 minutes), so you shouldn't have major issues with building the images and experimenting (or asking me for help). Just obviously, use `cmake-rs` or `meson-rs` with vcpkg or Conan, and not the raw commands, so it can properly integrate it with the Rust build system.
